[Core] Auto mapping for cluster name

Michaelvll opened the pull request:
<!-- Describe the changes in this PR -->

Closes #1853 and Closes #963 

The backward compatibility is guaranteed by the restore of ray yaml in the `backend_utils.write_cluster_config`


<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch -c i-am-a-very-long-cluster-name-with-more-than-35-characters --cloud gcp --cpus 2`
  - [x] `sky status -r`
  - [x] `sky exec  i-am-a-very-long-cluster-name-with-more-than-35-characters echo hi`
  - [x] `sky autostop -i  1 i-am-a-very-long-cluster-name-with-more-than-35-characters`
  - [x] `sky status -r`
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [x] All smoke tests: `pytest tests/test_smoke.py --aws` 
- [x] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

cblmemo commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2192,11 +2197,12 @@ class CloudVmRayResourceHandle(backends.backend.ResourceHandle):
     def __init__(self,
                  *,
                  cluster_name: str,
+                 cluster_name_on_cloud: str,
```
```suggestion
                 cluster_name_on_cloud: Optional[str],
```

Michaelvll replied:
The argument should not be None. The member variable will only be None for the clusters created before this PR.

cblmemo commented on sky/provision/aws/instance.py:
```diff
@@ -165,8 +164,7 @@ def cleanup_ports(
     # and backend_utils::write_cluster_config
-    sg_name = (f'sky-sg-{common_utils.user_and_hostname_hash()}'
-               f'-{common_utils.truncate_and_hash_cluster_name(cluster_name)}')
+    sg_name = f'sky-sg-{cluster_name}'
```
ditto

Michaelvll replied:
I changed this to read directly from teh `provider_config` for the security group name. Does that make sense?

cblmemo commented on sky/utils/common_utils.py:
```diff
@@ -88,11 +94,46 @@ def _is_valid_user_hash(user_hash: Optional[str]) -> bool:
+        user_hash = f'-{user_hash}'
+    user_hash_length = len(user_hash)
+
+    if (max_length is None or
```
add a doc string for meaning when `max_length is None`?

Michaelvll replied:
Good point! Added the docstr. Thanks!

cblmemo reviewed (commented):
Just briefly go through the code and it looks pretty good to me! Left some comments now and I'll read it again later ;)

One question to discuss: Shall we add a new tag like `sky-original-cluster-name` on cloud to let user find their corresponding machine easier?

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1082,6 +1086,9 @@ def write_cluster_config(
             f.write(restored_yaml_content)
 
+    yaml_config = common_utils.read_yaml(tmp_yaml_path)
+    config_dict['cluster_name_on_cloud'] = yaml_config['cluster_name']
```
It is possible that the yaml file is restored for backward compatibility from the previous section. We need to use the correct cluster name from that yaml file.

Michaelvll commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2179,7 +2182,9 @@ def provision_with_retries(
+
+    The handle object will last for the whole lifecycle of the cluster.
 
     - (required) Cluster name.
```
Added. Thanks!

Michaelvll commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2179,7 +2182,9 @@ def provision_with_retries(
 
 
 class CloudVmRayResourceHandle(backends.backend.ResourceHandle):
-    """A pickle-able tuple of:
```
Done. Thanks!

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -981,10 +982,13 @@ def write_cluster_config(
     default_aws_sg_name = f'sky-sg-{common_utils.user_and_hostname_hash()}'
     if ports is not None:
-        default_aws_sg_name += f'-{common_utils.truncate_and_hash_cluster_name(cluster_name)}'
+        default_aws_sg_name = f'sky-sg-{cluster_name_on_cloud}'
```
The `cluster_name_on_cloud` already contains the user hash, so it should be fine to don't have the `user_and_hostname_hash()`?

Michaelvll reviewed (commented):
> Just briefly go through the code and it looks pretty good to me! Left some comments now and I'll read it again later ;)
> 
> One question to discuss: Shall we add a new tag like `sky-original-cluster-name` on the cloud to let the user find their corresponding machine easier?

Thanks for the review @cblmemo! Added the tag for the original cluster name and addressed the comments. PTAL.

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -994,7 +998,8 @@ def write_cluster_config(
             resources_vars,
             **{
-                'cluster_name': cluster_name,
+                'cluster_name': cluster_name_on_cloud,
+                'user_cluster_name': cluster_name,
```
How about naming the two keys `cluster_name_on_cloud`, `cluster_name`, for consistency?

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1827,26 +1837,30 @@ def _query_cluster_status_via_cloud_api(
         cloud_name = repr(handle.launched_resources.cloud)
         try:
             node_status_dict = provision_lib.query_instances(
```
It may be pretty easy for new code/contributors to call provision_lib funcs and pass `cluster_name=cluster_name`. Hard to catch if we don't rename the args in provision_lib interface. Is this too intrusive? Any ideas to mitigate this?

One possibility/bandage is to document this clearly at the top of provision_lib.

concretevitamin replied:
Another potential way is to create a lightweight class and use it in provision_lib interfaces:
```
class ClusterIdentity:  # or better name
  cluster_name: str
  cluster_name_on_cloud: str
  cloud: str 
```

Michaelvll replied:
This is a great point! I could not think of a better way to mitigate this. I would personally prefer renaming the `cluster_name` in `provision_lib` to `cluster_name_on_cloud`, and add a comment at the top of `provision_lib`.

Another alternative (though more involved) is to rename the `handle.cluster_name` to `handle.display_name` and rename `handle.cluster_name_on_cloud` to `handle.cluster_name`.

Michaelvll replied:
After offline discussion, we now changed the `provision_lib`'s APIs to take `cluster_name_on_cloud` as an argument.

concretevitamin reviewed (commented):
Awesome @Michaelvll. Did a partial pass, leaving some comments first.

concretevitamin commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2179,9 +2182,14 @@ def provision_with_retries(
 
     - (required) Cluster name.
+    - (required) Cluster name on cloud (different from the cluster name, as we
+        append user hash to avoid confliction across multiple accounts in a same
+        organization, and truncate the name for length limit).
```
```suggestion
        append user hash to avoid conflict b/t multiple users in the same
        organization/account, and truncate the name for length limit).
```

concretevitamin commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -3896,7 +3925,7 @@ def post_teardown_cleanup(self,
                         _TEARDOWN_FAILURE_MESSAGE.format(
                             extra_reason='It is caused by TPU failure.',
-                            cluster_name=handle.cluster_name,
+                            cluster_name=cluster_name_on_cloud,
```
ditto

concretevitamin commented on tests/test_spot.py:
```diff
@@ -45,6 +45,7 @@ def _mock_cluster_state(self, _mock_db_conn):
         assert 'state.db' not in global_user_state._DB.db_path
         handle = backends.CloudVmRayResourceHandle(
             cluster_name='test-cluster1',
+            cluster_name_on_cloud='test-cluster1',
```
Should these be handled automatically by back-compat code?

cblmemo commented on sky/spot/spot_utils.py:
```diff
@@ -202,7 +203,14 @@ def callback_func(state: str):
+    # Truncate the task name to 30 chars to avoid the cluster name being too
+    # long after appending the job id, which will cause another truncation in
+    # the underlying sky.launch, hiding the `job_id` in the cluster name.
+    cluster_name = common_utils.truncate_and_hash_cluster_name(
```
nit: the variable is not `cluster_name` but `truncated_task_name`..?

cblmemo commented on sky/utils/common_utils.py:
```diff
@@ -64,6 +69,7 @@ def _is_valid_user_hash(user_hash: Optional[str]) -> bool:
             return False
         return len(user_hash) == USER_HASH_LENGTH
 
+    default_value = os.getenv(constants.USER_ID_ENV)
     user_hash = default_value
```
```suggestion
    user_hash = os.getenv(constants.USER_ID_ENV)
```

cblmemo commented on tests/test_smoke.py:
```diff
@@ -2046,6 +2046,8 @@ def test_spot_recovery_gcp():
 def test_spot_pipeline_recovery_aws(aws_config_region):
     """Test managed spot recovery for a pipeline."""
     name = _get_cluster_name()
+    user_hash = common_utils.get_user_hash()
+    user_hash = user_hash[:common_utils.USER_HASH_LENGTH_IN_CLUSTER_NAME]
```
nit: Shall we pack these two lines into a function?

cblmemo replied:
nit: Shall we pack these two lines into a function?

cblmemo commented on tests/test_smoke.py:
```diff
@@ -2089,8 +2094,11 @@ def test_spot_pipeline_recovery_gcp():
     zone = 'us-east4-b'
+    user_hash = common_utils.get_user_hash()
+    user_hash = user_hash[:common_utils.USER_HASH_LENGTH_IN_CLUSTER_NAME]
+    # TODO(zhwu): fix the name for spot cluster.
```
What is this TODO means?

cblmemo commented on tests/test_smoke.py:
```diff
@@ -2149,6 +2157,8 @@ def test_spot_recovery_default_resources(generic_cloud: str):
 def test_spot_recovery_multi_node_aws(aws_config_region):
     """Test managed spot recovery."""
     name = _get_cluster_name()
+    name_on_cloud = common_utils.truncate_and_hash_cluster_name(
+        name, spot.SPOT_CLUSTER_NAME_PREFIX_LENGTH, add_user_hash=False)
```
nit: pack these two lines into a function too? They appear a lot of times

cblmemo commented on tests/test_smoke.py:
Just to make sure, it seems we don't have any test that query based on cluster name other than three major clouds?

cblmemo replied:
Oh nvm... I was about to say it is a little bit strange that there is not something like
```bash
curl <lambda_api_addr> -X POST -d '{"cluster_name": "xxx"}'
```
in all changes inside `test_smoke.py`, from a CLI call that is
* not from three major cloud (e.g. lambda);
* contains some filter conditions including cluster name on the cloud.
But I just briefly went through all smoke tests and it seems like there are indeed no such tests 🤔

cblmemo reviewed (commented):
Thanks for the fix! Left several nits on top of Zongheng's thorough review :)

Michaelvll commented:
> Logging/UX: Tried sky spot launch --cpus 2+ -n spot-maskgit-minerl-maskgit-interp --cloud gcp 'for i in $(seq 10000); do echo $i; sleep 1; done' which is a long name a user told us before. In controller logs

Hey @concretevitamin, thanks for pointing this out! We intentionally truncate the cluster name earlier to 30 chars to avoid the cluster name being too long after appending the job id, which will cause another truncation in the underlying `sky.launch`, hiding the `job_id` in the cluster name. I think it should be fine, as a normal user is not expected to use `sky status` in the spot controller, and the cluster name in the controller log is not critical as there will be a single cluster name for each spot job. If we check the `sky spot queue`, the spot job should be the same as the user specified. Wdyt?

Michaelvll commented:
Thanks for the detailed review @concretevitamin @cblmemo! I fixed most of the comments.

I just realized that we should not tag clusters with original cluster names, as for most of the clouds, tags have the same length limit as cluster names, so I removed the tags.

TODO:
- [ ] Add a readme for design of cluster name mapping
- [ ] Test again.

cblmemo commented:
> Thanks for the detailed review @concretevitamin @cblmemo! I fixed most of the comments.
> 
> I just realized that we should not tag clusters with original cluster names, as for most of the clouds, tags have the same length limit as cluster names, so I removed the tags.
> 
> TODO:
> 
> * [ ]  Add a readme for design of cluster name mapping
> * [ ]  Test again.

Just a quick thought -  then maybe we could have a column for `cluster_name_on_cloud` for `sky status -a`..?

concretevitamin commented on sky/skylet/constants.py:
```diff
@@ -70,3 +70,7 @@
     # cause error and waiting for the error to be reported: #2273.
     'which conda | grep /opt/conda || conda init > /dev/null;')
+
+USER_ID_ENV_VAR = 'SKYPILOT_USER_ID'
+
+CLUSTER_NAME_VALID_REGEX = '[a-z]([-a-z0-9]*[a-z0-9])?'
```
Brief docstrs? (We should adopt this convention for all public members.)

concretevitamin commented on sky/design_docs/cluster_name.md:
```diff
@@ -0,0 +1,45 @@
+
+After #2403, SkyPilot will automatically truncate cluster names to comply with the
+naming length limit of the cloud, and append a user hash to the end of the name to
+distinguish clusters from different users.
```
```suggestion
distinguish clusters from different users (we make sure the final result after the append is within length limit).
```

concretevitamin commented on sky/design_docs/cluster_name.md:
```diff
@@ -0,0 +1,45 @@
+## Details
+
+Cluster names used in different situations:
+- CLI/API/UI: the name users assign to the cluster, i.e., the original cluster name
+    without any modification.
+- Lower-level components: for the components that directly interact with the cloud
+    provider, the newly generated name will be used, including, the underlying
+    `ray up`, `provision` module, etc.
```
Can we document the two VM tags?

concretevitamin replied:
```suggestion
- CLI/API/UI: the name users assign to the cluster, i.e., the original cluster name
    without any modification. `cluster_name` in code.
- Lower-level components: for the components that directly interact with the cloud
    provider, the newly generated name will be used, including, the underlying
    `ray up`, `provision` module, etc. `cluster_name_on_cloud` in code.
```

concretevitamin commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2179,9 +2182,14 @@ def provision_with_retries(
 
     - (required) Cluster name.
+    - (required) Cluster name on cloud (different from the cluster name, as we
+        append user hash to avoid conflict b/t multiple users in the same
+        organization/account, and truncate the name for length limit).
```
can replace w/ "See design_docs/cluster-name.md"?

concretevitamin commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2192,11 +2200,12 @@ class CloudVmRayResourceHandle(backends.backend.ResourceHandle):
     - (optional) If TPU(s) are managed, a path to a deletion script.
     """
-    _VERSION = 5
+    _VERSION = 6
```
`# Bump if any fields get added/removed/changed(?).`

concretevitamin commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -3377,8 +3391,8 @@ def _teardown(self,
                     f'identity mismatch error and removing '
-                    f'the cluser record from cluster table.{reset}\n{yellow}It '
-                    'is the user\'s responsibility to ensure that this '
+                    f'the cluster record from cluster table.{reset}\n{yellow}It'
```
Nice! I use Code Spell Checker in vscode which can flag all typos.

concretevitamin commented on sky/templates/gcp-ray.yml.j2:
```diff
@@ -63,6 +63,7 @@ available_node_types:
     node_config:
       labels:
         skypilot-user: {{ user }}
+        skypilot-cluster-name: {{ cluster_name }}
```
Just tried launching `sky launch --cpus 2+ -c dbg --use-spot --cloud gcp`. The VM has an empty value for this key.

Michaelvll replied:
Oops, this is a remnant, we decided to remove it from the tag, as the tag has the same character length limit

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -994,7 +999,7 @@ def write_cluster_config(
             resources_vars,
             **{
-                'cluster_name': cluster_name,
+                'cluster_name_on_cloud': cluster_name_on_cloud,
```
Need to pass 'cluster_name' too to avoid empty tags.

Michaelvll commented on sky/templates/azure-ray.yml.j2:
```diff
@@ -92,6 +92,7 @@ available_node_types:
       node_config:
         tags:
           skypilot-user: {{ user }}
+          skypilot-cluster-name: {{ cluster_name }}
```
Reimoved. Thanks!

Michaelvll commented on sky/templates/aws-ray.yml.j2:
Removed.

concretevitamin commented on sky/skylet/constants.py:
```diff
@@ -71,6 +71,11 @@
+# will be recognized as the same user (e.g., spot controller).
 USER_ID_ENV_VAR = 'SKYPILOT_USER_ID'
 
+# In most of clouds, cluster name can only contain lowercase letters, numbers
```
```suggestion
# In most clouds, cluster names can only contain lowercase letters, numbers
```

concretevitamin commented on sky/design_docs/cluster_name.md:
```diff
@@ -11,16 +11,19 @@ when the following situation happens, it can be annoying:
+    `ray up`, `provision` module, etc. `cluster_name_on_cloud` in code.
+- Cluster name on cloud: names of clusters on the cloud will always be
+    `cluster_name_on_cloud`, including the cluster name shown in cloud consoles,
+    the cluster name in the tag (ray-cluster-name), etc.
```
I'd suggest the following 2 items rather than 3 concepts:

--

There are two cluster name concepts:
- **SkyPilot cluster name**: This is the original "display" name users pick (or generated by the system) for each SkyPilot cluster. This is used in our CLI/API (e.g., `sky launch -c <cluster name>`). 
  - `cluster_name` in code. Used in all user-facing logging.
- **Cluster name on cloud**: This is the auto-generated name passed to the cloud. We auto-process the display name above to produce this new name to ensure (1) distinguishing across users, in case they choose the same cluster name; (2) the cluster name on cloud is within length limit.
  - `cluster_name_on_cloud` in code. Used in all modules that communicated with the cloud provider, e.g., `ray up`, node providers/`provision_lib` module, etc.
  - Recorded in the VM tag `ray-cluster-name`. (Note that the VM names include it as a substring, but may have additional prefixes/suffixes.)

In most cases, users need only be aware of the former.

Michaelvll replied:
Good point! Updated. Thanks!

concretevitamin reviewed (approved):
LGTM @Michaelvll - thanks for this major improvement!
